File: src/lib/python-backend/api/rateLimits.ts
/**
 * Rate Limits API Client
 * Frontend client for interacting with the Python backend rate limiting service.
 * 
 * Provides quota tracking and pre-publish validation for all social platforms.
 */
import { get, post } from '../client';
import { ENDPOINTS } from '../config';

// =============================================================================
// TYPES
// =============================================================================

export type Platform =
    | 'facebook'
    | 'instagram'
    | 'twitter'
    | 'linkedin'
    | 'tiktok'
    | 'youtube'
    | 'meta_ads';

export interface PlatformQuota {
    used: number;
    limit: number;
    remaining: number;
    percentage: number;
    isExceeded: boolean;
    isWarning: boolean;
    isCritical: boolean;
    resetsAt: string;
    description: string;
}

export interface AllQuotasResponse {
    success: boolean;
    quotas: Record<Platform, PlatformQuota>;
    summary: {
        totalPlatforms: number;
        exceededCount: number;
        warningCount: number;
        exceededPlatforms: string[];
        warningPlatforms: string[];
    };
    timestamp: string;
}

export interface QuotaCheckRequest {
    workspace_id: string;
    platform: string;
    post_count?: number;
}

export interface QuotaCheckResponse {
    success: boolean;
    allowed: boolean;
    platform: string;
    used: number;
    limit: number;
    remaining: number;
    message: string;
    error_code?: string;
}

export interface PlatformLimit {
    postsPerDay: number;
    apiCallsPerHour: number;
    apiCallsPerMinute: number;
    commentWritesPerHour: number;
    description: string;
    isDynamic: boolean;
}

export interface UsageHistoryItem {
    id: string;
    workspace_id: string;
    platform: string;
    date: string;
    posts_count: number;
    daily_limit: number;
    created_at: string;
    updated_at: string;
}

// =============================================================================
// API FUNCTIONS
// =============================================================================

/**
 * Get quota status for all platforms.
 * Use this for dashboard display.
 */
export async function getAllQuotas(workspaceId: string): Promise<AllQuotasResponse> {
    return get<AllQuotasResponse>(
        `${ENDPOINTS.rateLimits.status}?workspace_id=${workspaceId}`
    );
}

/**
 * Get quota status for a specific platform.
 */
export async function getPlatformQuota(
    platform: Platform,
    workspaceId: string
): Promise<{ success: boolean } & PlatformQuota> {
    return get<{ success: boolean } & PlatformQuota>(
        `${ENDPOINTS.rateLimits.platform(platform)}?workspace_id=${workspaceId}`
    );
}

/**
 * Check if publishing is allowed before attempting to post.
 * Call this BEFORE publishing to verify quota is available.
 */
export async function checkCanPublish(
    workspaceId: string,
    platform: string,
    postCount: number = 1
): Promise<QuotaCheckResponse> {
    return post<QuotaCheckResponse>(
        ENDPOINTS.rateLimits.check,
        {
            workspace_id: workspaceId,
            platform,
            post_count: postCount,
        }
    );
}

/**
 * Increment usage after successful publish.
 * Call this AFTER a post is successfully published.
 */
export async function incrementUsage(
    workspaceId: string,
    platform: string,
    count: number = 1
): Promise<{
    success: boolean;
    message: string;
    currentUsage: number;
    remaining: number;
    limit: number;
}> {
    return post(ENDPOINTS.rateLimits.increment, {
        workspace_id: workspaceId,
        platform,
        count,
    });
}

/**
 * Get usage history for analytics.
 */
export async function getUsageHistory(
    workspaceId: string,
    platform?: Platform,
    days: number = 7
): Promise<{
    success: boolean;
    history: UsageHistoryItem[];
    days: number;
    platform: string | null;
}> {
    let url = `${ENDPOINTS.rateLimits.history(workspaceId)}?days=${days}`;
    if (platform) {
        url += `&platform=${platform}`;
    }
    return get(url);
}

/**
 * Get configured limits for all platforms.
 * Returns official rate limits from platform documentation.
 */
export async function getPlatformLimits(): Promise<{
    success: boolean;
    limits: Record<Platform, PlatformLimit>;
    platforms: Platform[];
}> {
    return get(ENDPOINTS.rateLimits.limits);
}

// =============================================================================
// HELPER FUNCTIONS
// =============================================================================

/**
 * Check if user is approaching quota limit (80%+).
 */
export function isApproachingLimit(quota: PlatformQuota): boolean {
    return quota.isWarning && !quota.isExceeded;
}

/**
 * Get quota status color for UI display.
 */
export function getQuotaStatusColor(quota: PlatformQuota): 'green' | 'yellow' | 'red' {
    if (quota.isExceeded || quota.isCritical) return 'red';
    if (quota.isWarning) return 'yellow';
    return 'green';
}

/**
 * Format remaining quota for display.
 */
export function formatQuotaRemaining(quota: PlatformQuota): string {
    if (quota.isExceeded) {
        return 'Limit reached';
    }
    return `${quota.remaining} remaining`;
}

/**
 * Get time until quota resets.
 */
export function getTimeUntilReset(resetsAt: string): string {
    const resetTime = new Date(resetsAt);
    const now = new Date();
    const diffMs = resetTime.getTime() - now.getTime();

    if (diffMs <= 0) return 'Resetting soon';

    const hours = Math.floor(diffMs / (1000 * 60 * 60));
    const minutes = Math.floor((diffMs % (1000 * 60 * 60)) / (1000 * 60));

    if (hours > 0) {
        return `Resets in ${hours}h ${minutes}m`;
    }
    return `Resets in ${minutes}m`;
}

File: python_backend/src/api/v1/rate_limits.py
"""
Rate Limits API Routes
Endpoints for quota management and usage tracking

Endpoints:
- GET /api/v1/rate-limits/status - Get all platform quotas
- GET /api/v1/rate-limits/{platform} - Get single platform quota
- POST /api/v1/rate-limits/check - Pre-publish quota check
- POST /api/v1/rate-limits/cleanup - Admin: cleanup old records (cron)
"""
import logging
from typing import Optional
from fastapi import APIRouter, HTTPException, Request, Query
from pydantic import BaseModel

from ...services.rate_limit_service import (
    RateLimitService,
    get_rate_limit_service
)
from ...services.rate_limit_constants import Platform, PLATFORM_LIMITS

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/rate-limits", tags=["Rate Limits"])


# =============================================================================
# REQUEST/RESPONSE MODELS
# =============================================================================

class QuotaCheckRequest(BaseModel):
    """Request model for quota check"""
    workspace_id: str
    platform: str
    post_count: int = 1


class QuotaCheckResponse(BaseModel):
    """Response model for quota check"""
    success: bool
    allowed: bool
    platform: str
    used: int
    limit: int
    remaining: int
    message: str
    error_code: Optional[str] = None


class IncrementUsageRequest(BaseModel):
    """Request model for incrementing usage"""
    workspace_id: str
    platform: str
    count: int = 1


# =============================================================================
# ENDPOINTS
# =============================================================================

@router.get("/status")
async def get_all_quotas(
    workspace_id: str = Query(..., description="Workspace ID")
):
    """
    Get quota status for all platforms.
    
    Returns usage, limits, and warnings for each platform.
    Use this for dashboard display.
    """
    try:
        service = get_rate_limit_service()
        result = await service.get_all_quotas(workspace_id)
        return {"success": True, **result}
    except Exception as e:
        logger.error(f"Get quotas error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/{platform}")
async def get_platform_quota(
    platform: str,
    workspace_id: str = Query(..., description="Workspace ID")
):
    """
    Get quota status for a specific platform.
    
    Args:
        platform: Platform name (facebook, instagram, twitter, linkedin, tiktok, youtube)
        workspace_id: Workspace ID
    """
    try:
        # Validate platform
        valid_platforms = [p.value for p in Platform]
        if platform.lower() not in valid_platforms:
            raise HTTPException(
                status_code=400,
                detail=f"Invalid platform. Must be one of: {', '.join(valid_platforms)}"
            )
        
        service = get_rate_limit_service()
        status = await service.get_quota_status(workspace_id, platform)
        
        return {
            "success": True,
            "platform": status.platform,
            "used": status.used,
            "limit": status.limit,
            "remaining": status.remaining,
            "percentage": status.percentage,
            "isExceeded": status.is_exceeded,
            "isWarning": status.is_warning,
            "isCritical": status.is_critical,
            "resetsAt": status.resets_at,
            "description": status.description
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get platform quota error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/check")
async def check_quota(request: QuotaCheckRequest) -> QuotaCheckResponse:
    """
    Pre-publish quota check.
    
    Call this BEFORE attempting to publish to verify quota is available.
    Returns whether publishing is allowed and quota details.
    """
    try:
        service = get_rate_limit_service()
        result = await service.check_can_publish(
            workspace_id=request.workspace_id,
            platform=request.platform,
            post_count=request.post_count
        )
        
        return QuotaCheckResponse(
            success=True,
            allowed=result.allowed,
            platform=result.platform,
            used=result.used,
            limit=result.limit,
            remaining=result.remaining,
            message=result.message,
            error_code=result.error_code
        )
    except Exception as e:
        logger.error(f"Check quota error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/increment")
async def increment_usage(request: IncrementUsageRequest):
    """
    Increment usage after successful publish.
    
    Call this AFTER a post is successfully published.
    This is typically called by the social service internally.
    """
    try:
        service = get_rate_limit_service()
        success = await service.increment_usage(
            workspace_id=request.workspace_id,
            platform=request.platform,
            count=request.count
        )
        
        if success:
            # Return updated quota status
            status = await service.get_quota_status(request.workspace_id, request.platform)
            return {
                "success": True,
                "message": f"Usage incremented by {request.count}",
                "currentUsage": status.used,
                "remaining": status.remaining,
                "limit": status.limit
            }
        else:
            raise HTTPException(status_code=500, detail="Failed to increment usage")
            
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Increment usage error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/history/{workspace_id}")
async def get_usage_history(
    workspace_id: str,
    platform: Optional[str] = None,
    days: int = Query(default=7, ge=1, le=30)
):
    """
    Get usage history for analytics.
    
    Args:
        workspace_id: Workspace ID
        platform: Optional platform filter
        days: Number of days to fetch (1-30, default: 7)
    """
    try:
        service = get_rate_limit_service()
        history = await service.get_usage_history(
            workspace_id=workspace_id,
            platform=platform,
            days=days
        )
        
        return {
            "success": True,
            "history": history,
            "days": days,
            "platform": platform
        }
    except Exception as e:
        logger.error(f"Get history error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/cleanup")
async def cleanup_old_records(request: Request):
    """
    Admin endpoint: Clean up old rate limit records.
    
    Should be called by cron job. Deletes records older than 30 days.
    Requires CRON_SECRET header for authentication.
    """
    try:
        # Verify cron secret
        cron_secret = request.headers.get("x-cron-secret")
        import os
        expected_secret = os.getenv("CRON_SECRET", "")
        
        if expected_secret and cron_secret != expected_secret:
            raise HTTPException(status_code=401, detail="Invalid cron secret")
        
        service = get_rate_limit_service()
        result = await service.reset_daily_quotas()
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Cleanup error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/limits")
async def get_platform_limits():
    """
    Get configured limits for all platforms.
    
    Returns the official rate limits based on platform documentation.
    Useful for displaying limits in the UI.
    """
    try:
        limits = {}
        for platform in Platform:
            limit_config = PLATFORM_LIMITS.get(platform.value)
            if limit_config:
                limits[platform.value] = {
                    "postsPerDay": limit_config.posts_per_day,
                    "apiCallsPerHour": limit_config.api_calls_per_hour,
                    "apiCallsPerMinute": limit_config.api_calls_per_minute,
                    "commentWritesPerHour": limit_config.comment_writes_per_hour,
                    "description": limit_config.description,
                    "isDynamic": limit_config.is_dynamic
                }
        
        return {
            "success": True,
            "limits": limits,
            "platforms": [p.value for p in Platform]
        }
    except Exception as e:
        logger.error(f"Get limits error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/")
async def rate_limits_info():
    """Rate Limits API information"""
    return {
        "success": True,
        "message": "Rate Limits API is operational",
        "version": "1.0.0",
        "endpoints": {
            "status": "GET /status - Get all platform quotas",
            "platform": "GET /{platform} - Get single platform quota",
            "check": "POST /check - Pre-publish quota check",
            "increment": "POST /increment - Record successful publish",
            "history": "GET /history/{workspace_id} - Usage history",
            "limits": "GET /limits - Get configured limits",
            "cleanup": "POST /cleanup - Admin cleanup (cron)"
        }
    }
